File: packages/scelo-r/R/io.R
# Loading and saving: the soft-data intake, with Scelo's typing rules.
#
# sc_load(path) reads CSV / TSV / TXT (delimiter sniffed) / RDS and applies
# the same cell coercion Scelo IDE applies at import (coerceCsvCell in
# packages/scelo-core): a small set of missing tokens becomes NA, only
# strictly numeric strings become numbers, id-like integers ("007", > 2^53)
# stay strings. Nothing else is touched: cleaning is an explicit, auditable
# step (sc_clean), not something that happens on the way in.

SC_MISSING_CELL_TOKENS <- c("null", "na", "n/a", "nan", "none", "-")
.SC_NUMERIC_STRING_RE <- "^[+-]?([0-9]+(\\.[0-9]*)?|\\.[0-9]+)([eE][+-]?[0-9]+)?$"

#' Scelo's import-time cell rule
#'
#' `""` and the small missing-token set become `NA`; strictly numeric strings
#' become numbers; leading-zero integers ("007") and integers beyond 2^53
#' stay strings; everything else is returned as is.
#' @param x A character vector.
#' @return A list of coerced cells (use [sc_coerce()] for a whole frame).
#' @examples
#' sc_coerce_cell(c(" 42 ", "007", "NA", "1e3", "0x1f"))
#' @export
sc_coerce_cell <- function(x) {
  s <- trimws(as.character(x))
  out <- as.list(s)
  miss <- is.na(s) | s == "" | (nchar(s) <= 4 & tolower(s) %in% SC_MISSING_CELL_TOKENS)
  num <- !miss & grepl(.SC_NUMERIC_STRING_RE, s)
  plain_int <- num & grepl("^[+-]?[0-9]+$", s)
  leading0 <- plain_int & grepl("^[+-]?0[0-9]", s)
  vals <- suppressWarnings(as.numeric(s))
  big <- plain_int & !is.na(vals) & abs(vals) > 2^53 - 1
  keep_str <- leading0 | big
  for (i in which(miss)) out[[i]] <- NA
  for (i in which(num & !keep_str & is.finite(vals))) out[[i]] <- vals[i]
  out
}

.sc_is_text <- function(col) is.character(col) || is.factor(col)

.sc_coerce_column <- function(col) {
  s <- as.character(col)
  cells <- sc_coerce_cell(s)
  is_num <- vapply(cells, is.numeric, logical(1))
  is_na <- vapply(cells, function(v) length(v) == 1 && is.na(v), logical(1))
  if (all(is_na)) return(rep(NA_character_, length(cells)))
  if (all(is_num | is_na)) {
    v <- vapply(cells, function(c) if (is.numeric(c)) c else NA_real_, numeric(1))
    if (all(is.na(v) | v == round(v)) && !any(is.na(v)) && all(abs(v) < .Machine$integer.max)) return(as.integer(v))
    return(v)
  }
  # mixed: keep the numeric cells as numbers inside a list? R columns are typed; keep character but
  # with the coerced strings (trimmed) so the cleaning layer can type it (coerce-numeric).
  vapply(cells, function(c) if (length(c) == 1 && is.na(c)) NA_character_ else as.character(c), character(1))
}

#' Apply Scelo's import coercion to every text column
#'
#' @param df A data frame.
#' @return A data frame with numeric-looking text columns typed as numbers.
#' @export
sc_coerce <- function(df) {
  for (c in names(df)) if (.sc_is_text(df[[c]])) df[[c]] <- .sc_coerce_column(df[[c]])
  df
}

#' Guess a delimiter the way the IDE does
#'
#' Reads the first KB, rejects binary content (NUL bytes, > 2 % control
#' bytes), drops the last (possibly truncated) line, keeps 20 lines and picks
#' the candidate among comma, tab and semicolon whose minimum per-line count is highest.
#' @param path A file.
#' @param nbytes Bytes to inspect.
#' @return A one-character delimiter (comma, tab or semicolon) or `NULL`.
#' @export
sc_sniff <- function(path, nbytes = 1024L) {
  raw <- readBin(path, "raw", nbytes)
  if (!length(raw)) return(NULL)
  b <- as.integer(raw)
  if (any(b == 0L)) return(NULL)
  control <- sum(b < 32L & !(b %in% c(9L, 10L, 13L)))
  if (control / length(b) > 0.02) return(NULL)
  text <- rawToChar(raw[b != 0L])
  Encoding(text) <- "UTF-8"
  lines <- strsplit(text, "\r?\n")[[1]]
  if (length(lines) > 1) lines <- lines[-length(lines)]
  lines <- utils::head(lines[nzchar(trimws(lines))], 20)
  if (!length(lines)) return(NULL)
  best <- NULL
  best_min <- 0L
  for (cand in c(",", "\t", ";")) {
    m <- min(vapply(lines, function(l) lengths(regmatches(l, gregexpr(cand, l, fixed = TRUE))), integer(1)))
    if (m >= 1L && m > best_min) {
      best <- cand
      best_min <- m
    }
  }
  best
}

.sc_dedupe_header <- function(cols) {
  cols <- trimws(cols)
  cols[cols == ""] <- "column"
  out <- cols
  seen <- list()
  for (i in seq_along(cols)) {
    n <- cols[i]
    if (!is.null(seen[[n]])) {
      seen[[n]] <- seen[[n]] + 1L
      out[i] <- paste0(n, "_", seen[[n]])
    } else {
      seen[[n]] <- 1L
    }
  }
  out
}

#' Read a file into a data frame with Scelo's import typing
#'
#' Formats by extension: `.csv` `.tsv` `.txt` (delimiter sniffed) `.rds`.
#' Text is read as character and then typed cell by cell with
#' [sc_coerce_cell()]; headers are trimmed and de-duplicated (`_2`, `_3`).
#' @param path A file.
#' @param rows Keep a uniform reservoir sample of this many rows (the IDE
#'   caps imports at 250 000).
#' @param sep Delimiter override.
#' @param coerce Apply [sc_coerce()] (default `TRUE`).
#' @param seed Seed for the reservoir sample.
#' @param ... Passed to [utils::read.table()].
#' @return A data frame with attributes `name` and `source`.
#' @examples
#' df <- sc_load(system.file("extdata", "claims.csv", package = "scelo"))
#' head(df)
#' @export
sc_load <- function(path, rows = NULL, sep = NULL, coerce = TRUE, seed = NULL, ...) {
  if (!file.exists(path)) stop(sprintf("%s does not exist", path), call. = FALSE)
  ext <- tolower(tools::file_ext(path))
  if (ext == "rds") {
    df <- readRDS(path)
  } else {
    if (is.null(sep)) sep <- switch(ext, tsv = "\t", csv = ",", sc_sniff(path))
    if (is.null(sep)) stop(sprintf("%s does not look like delimited text (binary content, or no consistent delimiter in the first KB): try .csv, .tsv", basename(path)), call. = FALSE)
    df <- utils::read.table(path, sep = sep, header = TRUE, colClasses = "character", na.strings = character(), quote = "\"", comment.char = "",
                            check.names = FALSE, strip.white = FALSE, blank.lines.skip = TRUE, encoding = "UTF-8", fileEncoding = "UTF-8-BOM", stringsAsFactors = FALSE, ...)
    names(df) <- .sc_dedupe_header(names(df))
  }
  if (coerce) df <- sc_coerce(df)
  if (!is.null(rows) && nrow(df) > rows) df <- sc_reservoir(df, rows, seed = seed)
  attr(df, "name") <- basename(path)
  attr(df, "source") <- normalizePath(path)
  .sc_record("sc_load", list(path = path), NULL, df, 0)
  df
}

#' Uniform reservoir sample in original order
#'
#' The IDE's import cap: a uniform sample of `n` rows that keeps the file
#' order, stamped with `sampled`, `sample_kind` and `source_total_rows`.
#' @param df A data frame.
#' @param n Rows to keep.
#' @param seed Seed.
#' @return A data frame.
#' @export
sc_reservoir <- function(df, n, seed = NULL) {
  if (nrow(df) <= n) return(df)
  if (!is.null(seed)) set.seed(seed)
  idx <- sort(sample.int(nrow(df), n))
  out <- df[idx, , drop = FALSE]
  rownames(out) <- NULL
  attr(out, "sampled") <- TRUE
  attr(out, "sample_kind") <- "uniform"
  attr(out, "source_total_rows") <- nrow(df)
  out
}

#' Write a frame by extension
#'
#' `.csv` `.tsv` `.rds` `.json` (needs jsonlite) `.md` `.html`; notes and
#' basis of a `scelo_table` go into `.md` / `.html`. Writes atomically
#' (`<file>.partial` then rename).
#' @param df A data frame or `scelo_table`.
#' @param path Destination.
#' @param ... Passed to the writer.
#' @return The path, invisibly.
#' @export
sc_save <- function(df, path, ...) {
  ext <- tolower(tools::file_ext(path))
  dir.create(dirname(path), showWarnings = FALSE, recursive = TRUE)
  tmp <- paste0(path, ".partial")
  plain <- if (inherits(df, "scelo_table")) sc_df(df) else as.data.frame(df)
  if (ext %in% c("csv", "tsv", "txt")) {
    utils::write.table(plain, tmp, sep = if (ext == "tsv") "\t" else ",", row.names = FALSE, qmethod = "double", na = "", ...)
  } else if (ext == "rds") {
    saveRDS(df, tmp, ...)
  } else if (ext == "json") {
    .sc_need("jsonlite")
    writeLines(jsonlite::toJSON(plain, dataframe = "rows", pretty = TRUE, na = "null", digits = NA), tmp)
  } else if (ext == "md") {
    writeLines(sc_markdown(df), tmp)
  } else if (ext %in% c("html", "htm")) {
    writeLines(.sc_md_to_html(sc_markdown(df)), tmp)
  } else {
    stop(sprintf("unsupported extension '%s'", ext), call. = FALSE)
  }
  file.rename(tmp, path)
  invisible(path)
}

.sc_need <- function(pkg, why = NULL) {
  if (!requireNamespace(pkg, quietly = TRUE)) stop(sprintf("package '%s' is needed%s: install.packages(\"%s\")", pkg, if (is.null(why)) "" else paste0(" ", why), pkg), call. = FALSE)
  invisible(TRUE)
}

.SC_SAMPLES <- data.frame(
  key = c("claims", "climate", "dirty", "wmtr-scenarios", "lifelib-mp", "workspace-demo"),
  title = c("Synthetic claims", "Climate reanalysis ensemble", "Messy intake (dirty demo)", "WMTR forecast scenarios", "Lifelib model points", "Workspace demo"),
  about = c(
    "P&C reserving / pricing demo: 79 rows of an incomplete claims triangle (origins 2018-2024) with policy, line, province, age, sex, paid, incurred, settled.",
    "30 daily records for one grid cell (Pretoria, Jan 2024): 2-m temperature and precipitation under ERA5 / MERRA-2 / JRA-3Q.",
    "53-row customer ledger with every real-world mess: currency strings, %-numbers, sentinel ages, mixed booleans and date formats, mojibake, NBSP / zero-width characters, missing markers, duplicate rows.",
    "12 scenario rows for the W(M, T, R) Monte Carlo forecast: alpha_m / alpha_t / alpha_r, relational weights, shock, horizon.",
    "100-row in-force term-life model-point file shaped like lifelib's basic_term_sample.",
    "2,000-policy synthetic annuity book: three low-variance real drivers acting through nonlinear channels, a crude-rate level, ten high-variance nuisance columns."
  ),
  stringsAsFactors = FALSE
)

#' The bundled sample datasets
#'
#' The same six samples Scelo IDE offers, byte-identical.
#' @return A data frame of key, title, about.
#' @export
sc_samples <- function() .SC_SAMPLES

#' Load a bundled sample
#'
#' @param key One of `claims`, `climate`, `dirty`, `wmtr-scenarios`,
#'   `lifelib-mp`, `workspace-demo`.
#' @param ... Passed to [sc_load()].
#' @return A data frame.
#' @examples
#' head(sc_sample("claims"))
#' @export
sc_sample <- function(key = "claims", ...) {
  k <- gsub("_", "-", tolower(trimws(key)))
  if (!k %in% .SC_SAMPLES$key) stop(sprintf("unknown sample '%s': choose from %s", key, paste(.SC_SAMPLES$key, collapse = ", ")), call. = FALSE)
  df <- sc_load(.sc_extdata(paste0(k, ".csv")), ...)
  attr(df, "name") <- .SC_SAMPLES$title[.SC_SAMPLES$key == k]
  df
}

# Path of a bundled data file: the installed package, or inst/extdata when the
# sources are being run directly (dev/run-tests.R).
.sc_extdata <- function(file) {
  p <- system.file("extdata", file, package = "scelo")
  if (nzchar(p)) return(p)
  for (cand in c(file.path("inst", "extdata", file), file.path("..", "..", "inst", "extdata", file))) if (file.exists(cand)) return(normalizePath(cand))
  stop(sprintf("bundled file %s not found", file), call. = FALSE)
}

File: packages/scelo-py/src/scelo/io.py
"""Loading and saving: the soft-data intake, with Scelo's typing rules.

``load(path)`` reads CSV / TSV / TXT (delimiter sniffed) / Parquet / Excel /
JSON / Feather and applies the same cell coercion Scelo IDE applies at import
(``coerceCsvCell`` in packages/scelo-core): a small set of missing tokens
becomes null, only strictly numeric strings become numbers, id-like integers
("007", > 2^53) stay strings. Nothing else is touched: cleaning is an
explicit, auditable step (``scelo.clean``), not something that happens on
the way in.

Row cap: the IDE keeps a uniform reservoir sample past 250 000 rows so a
2-million-row extract cannot kill the window. ``load`` has no cap by default
(a script has the whole machine) but ``rows=`` gives you the same reservoir.
"""

from __future__ import annotations

import io as _io
import os
import re
from pathlib import Path
from typing import Any, Iterable, List, Optional, Union

import numpy as np
import pandas as pd

from ._audit import record
from ._table import Table

__all__ = [
    "load", "save", "coerce", "coerce_cell", "sniff", "samples", "sample", "reservoir",
    "DEFAULT_IMPORT_ROW_CAP", "MISSING_CELL_TOKENS",
]

PathLike = Union[str, "os.PathLike[str]"]

# Mirrors scelo-core: deliberately small and unambiguous. The long tail
# ("?", "TBD", "#N/A" …) is the cleaning layer's missing-tokens op.
MISSING_CELL_TOKENS = frozenset({"null", "na", "n/a", "nan", "none", "-"})
_NUMERIC_STRING_RE = re.compile(r"^[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?$")
_PLAIN_INTEGER_RE = re.compile(r"^[+-]?\d+$")
_LEADING_ZERO_RE = re.compile(r"^[+-]?0\d")
_SAFE_INT = 2**53 - 1

DEFAULT_IMPORT_ROW_CAP = 250_000

_DATA_DIR = Path(__file__).parent / "data"
_SAMPLES = {
    "claims": ("Synthetic claims", "P&C reserving / pricing demo: 79 rows of an incomplete claims triangle (origins 2018–2024) with policy, line, province, age, sex, paid, incurred, settled."),
    "climate": ("Climate reanalysis ensemble", "30 daily records for one grid cell (Pretoria, Jan 2024): 2-m temperature and precipitation under ERA5 / MERRA-2 / JRA-3Q."),
    "dirty": ("Messy intake (dirty demo)", "53-row customer ledger with every real-world mess: currency strings, %-numbers, sentinel ages, mixed booleans and date formats, mojibake, NBSP / zero-width characters, missing markers, duplicate rows."),
    "wmtr-scenarios": ("WMTR forecast scenarios", "12 scenario rows for the W(M, T, R) Monte Carlo forecast: alpha_m / alpha_t / alpha_r, relational weights, shock, horizon."),
    "lifelib-mp": ("Lifelib model points", "100-row in-force term-life model-point file shaped like lifelib's basic_term_sample (age_at_entry, sex, sum_assured, policy_term, duration_mth, premium_pp)."),
    "workspace-demo": ("Workspace demo", "2,000-policy synthetic annuity book: three low-variance real drivers acting through nonlinear channels, a crude-rate level, ten high-variance nuisance columns."),
}


# ── cell coercion ───────────────────────────────────────────────────────────

def coerce_cell(raw: Any) -> Any:
    """Scelo's import-time cell rule: "" / small missing tokens → None, strict numerics → number, else the string.

    >>> coerce_cell(" 42 "), coerce_cell("007"), coerce_cell("NA"), coerce_cell("1e3"), coerce_cell("0x1f")
    (42, '007', None, 1000.0, '0x1f')
    """
    if raw is None:
        return None
    if isinstance(raw, float):
        return None if np.isnan(raw) else raw
    if not isinstance(raw, str):
        return raw
    s = raw.strip()
    if s == "":
        return None
    if len(s) <= 4 and s.lower() in MISSING_CELL_TOKENS:
        return None
    if not _NUMERIC_STRING_RE.match(s):
        return s
    if _PLAIN_INTEGER_RE.match(s):
        if _LEADING_ZERO_RE.match(s):
            return s
        n = int(s)
        return n if abs(n) <= _SAFE_INT else s
    try:
        f = float(s)
    except ValueError:
        return s
    return f if np.isfinite(f) else s


def _coerce_object_column(col: pd.Series) -> pd.Series:
    """Apply :func:`coerce_cell` to an object column and settle its dtype.

    All-numeric → numeric dtype; ≥ 80 % numeric (Scelo's typing rule) → the
    column is left as object so the non-numeric residue ("6+") stays visible
    for the cleaning layer's coerce-numeric op; otherwise strings.
    """
    vals = col.astype(object).map(coerce_cell)
    vals = vals.astype(object).where(vals.notna(), None)
    nonnull = vals[vals.notna()]
    if len(nonnull) == 0:
        return vals
    is_num = nonnull.map(lambda v: isinstance(v, (int, float)) and not isinstance(v, bool))
    if is_num.all():
        out = pd.to_numeric(vals, errors="coerce")
        if (out.dropna() % 1 == 0).all() and not out.isna().any():
            return out.astype("int64")
        return out
    return vals


def coerce(df: pd.DataFrame) -> pd.DataFrame:
    """Apply Scelo's import coercion to every object column of a frame (returns a new frame)."""
    out = df.copy()
    for c in out.columns:
        if is_text(out[c]):
            out[c] = _coerce_object_column(out[c].astype(object))
    return out


def is_text(col: pd.Series) -> bool:
    """True for object / string-dtype columns (pandas ≥ 3 reads text as ``str`` dtype, older as ``object``)."""
    return col.dtype == object or pd.api.types.is_string_dtype(col)


# ── delimiter sniffing ──────────────────────────────────────────────────────

def sniff(path: PathLike, nbytes: int = 1024) -> Optional[str]:
    """Guess the delimiter of a text file the way the IDE does; None when it does not look like delimited text.

    Reads the first KB, rejects binary (NUL bytes, > 2 % control bytes),
    drops the last (possibly truncated) line, keeps 20 lines and picks the
    candidate in ``, \\t ;`` whose *minimum* per-line count is highest.
    """
    with open(path, "rb") as fh:
        head = fh.read(nbytes)
    if b"\x00" in head:
        return None
    control = sum(1 for b in head if b < 32 and b not in (9, 10, 13))
    if head and control / len(head) > 0.02:
        return None
    text = head.decode("utf-8", errors="replace")
    lines = re.split(r"\r?\n", text)
    if len(lines) > 1:
        lines = lines[:-1]
    lines = [ln for ln in lines if ln.strip()][:20]
    if not lines:
        return None
    best, best_min = None, 0
    for cand in (",", "\t", ";"):
        m = min(ln.count(cand) for ln in lines)
        if m >= 1 and m > best_min:
            best, best_min = cand, m
    return best


# ── reservoir sampling ──────────────────────────────────────────────────────

def reservoir(df: pd.DataFrame, n: int, seed: Optional[int] = None) -> pd.DataFrame:
    """Uniform sample of ``n`` rows in original order (the IDE's import cap), with provenance in ``attrs``."""
    if len(df) <= n:
        return df
    rng = np.random.default_rng(seed)
    idx = np.sort(rng.choice(len(df), size=n, replace=False))
    out = df.iloc[idx].reset_index(drop=True)
    out.attrs["sampled"] = True
    out.attrs["sample_kind"] = "uniform"
    out.attrs["source_total_rows"] = len(df)
    return out


# ── load / save ─────────────────────────────────────────────────────────────

def load(
    path: PathLike,
    *,
    rows: Optional[int] = None,
    sheet: Union[int, str, None] = 0,
    sep: Optional[str] = None,
    coerce_cells: bool = True,
    seed: Optional[int] = None,
    **kwargs: Any,
) -> pd.DataFrame:
    """Read a file into a DataFrame with Scelo's import typing.

    Formats by extension: ``.csv`` ``.tsv`` ``.txt`` (delimiter sniffed)
    ``.parquet`` ``.feather`` ``.xlsx`` ``.xls`` ``.json`` ``.jsonl``.

    Parameters
    ----------
    rows : keep a uniform reservoir sample of this many rows (the IDE uses 250 000).
    sheet : Excel sheet (index or name).
    sep : delimiter override for text files.
    coerce_cells : apply :func:`coerce_cell` to object columns (default True).
    kwargs : passed to the underlying pandas reader.
    """
    p = Path(path)
    if not p.exists():
        raise FileNotFoundError(f"{p} does not exist")
    ext = p.suffix.lower()
    if ext == ".parquet":
        df = pd.read_parquet(p, **kwargs)
    elif ext == ".feather":
        df = pd.read_feather(p, **kwargs)
    elif ext in (".xlsx", ".xls", ".xlsm"):
        df = pd.read_excel(p, sheet_name=sheet, **kwargs)
    elif ext == ".json":
        df = pd.read_json(p, **kwargs)
    elif ext == ".jsonl":
        df = pd.read_json(p, lines=True, **kwargs)
    else:
        if sep is None:
            sep = "\t" if ext == ".tsv" else ("," if ext == ".csv" else sniff(p))
            if sep is None:
                raise ValueError(
                    f"{p.name} does not look like delimited text (binary content, or no consistent delimiter in the first KB): try .csv, .tsv or .parquet"
                )
        kwargs.setdefault("dtype", str)
        kwargs.setdefault("keep_default_na", False)
        kwargs.setdefault("na_filter", False)
        kwargs.setdefault("skip_blank_lines", True)
        kwargs.setdefault("encoding", "utf-8-sig")
        df = pd.read_csv(p, sep=sep, **kwargs)
        df.columns = _dedupe_header([str(c).strip() for c in df.columns])
    if coerce_cells:
        df = coerce(df)
    if rows is not None and len(df) > rows:
        df = reservoir(df, rows, seed=seed)
    df.attrs.setdefault("name", p.name)
    df.attrs["source"] = str(p)
    record("load", {"path": str(p), "rows": rows}, None, df, 0.0)
    return df


def _dedupe_header(cols: List[str]) -> List[str]:
    """Empty names become ``column``; duplicates get ``_2``, ``_3`` …"""
    seen: dict = {}
    out: List[str] = []
    for c in cols:
        name = c if c else "column"
        if name in seen:
            seen[name] += 1
            name = f"{name}_{seen[name]}"
        else:
            seen[name] = 1
        out.append(name)
    return out


def save(df: pd.DataFrame, path: PathLike, **kwargs: Any) -> Path:
    """Write a frame by extension (.csv .tsv .parquet .feather .xlsx .json .jsonl .md .html); returns the path.

    CSV is written RFC-4180 (LF line endings, no index). A :class:`Table`'s
    notes and basis go into ``.md`` / ``.html`` output automatically.
    """
    p = Path(path)
    ext = p.suffix.lower()
    p.parent.mkdir(parents=True, exist_ok=True)
    tmp = p.with_name(p.name + ".partial")
    if ext in (".csv", ".tsv", ".txt"):
        kwargs.setdefault("index", False)
        kwargs.setdefault("lineterminator", "\n")
        pd.DataFrame(df).to_csv(tmp, sep="\t" if ext == ".tsv" else ",", **kwargs)
    elif ext == ".parquet":
        pd.DataFrame(df).to_parquet(tmp, index=False, **kwargs)
    elif ext == ".feather":
        pd.DataFrame(df).reset_index(drop=True).to_feather(tmp, **kwargs)
    elif ext in (".xlsx", ".xlsm"):
        pd.DataFrame(df).to_excel(tmp, index=False, **kwargs)
    elif ext == ".json":
        pd.DataFrame(df).to_json(tmp, orient="records", indent=2, **kwargs)
    elif ext == ".jsonl":
        pd.DataFrame(df).to_json(tmp, orient="records", lines=True, **kwargs)
    elif ext == ".md":
        text = df.to_markdown_report() if isinstance(df, Table) else _md(df)
        tmp.write_text(text + "\n", encoding="utf-8")
    elif ext in (".html", ".htm"):
        body = df._repr_html_() if isinstance(df, Table) else pd.DataFrame(df).to_html(index=False)
        tmp.write_text(body or "", encoding="utf-8")
    else:
        raise ValueError(f"unsupported extension {ext!r}")
    os.replace(tmp, p)  # atomic: <file>.partial → <file>, the house rule
    return p


def _md(df: pd.DataFrame) -> str:
    try:
        return pd.DataFrame(df).to_markdown(index=False)
    except ImportError:
        return "```\n" + pd.DataFrame(df).to_string(index=False) + "\n```"


# ── bundled samples ─────────────────────────────────────────────────────────

def samples() -> pd.DataFrame:
    """The bundled sample datasets (the same six Scelo IDE offers), as a table."""
    rows = [{"key": k, "title": t, "about": a} for k, (t, a) in _SAMPLES.items()]
    return pd.DataFrame(rows)


def sample(key: str = "claims", **kwargs: Any) -> pd.DataFrame:
    """Load a bundled sample: ``claims`` ``climate`` ``dirty`` ``wmtr-scenarios`` ``lifelib-mp`` ``workspace-demo``."""
    k = key.strip().lower().replace("_", "-")
    if k not in _SAMPLES:
        raise KeyError(f"unknown sample {key!r}: choose from {', '.join(_SAMPLES)}")
    df = load(_DATA_DIR / f"{k}.csv", **kwargs)
    df.attrs["name"] = _SAMPLES[k][0]
    return df
